yksgi/src/lib.rs

```diff
@@ -109,32 +121,53 @@ impl SGInterp {
     /// Create a new stopgap interpreter and initialise it to start interpretation at the location
     /// given by a basic block index, instruction index, and function name.
     /// FIXME: Support initialisation of multiple frames.
-    pub unsafe fn new(bbidx: usize, instridx: usize, fname: &CStr) -> SGInterp {
+    pub unsafe fn new(activeframes: &[FrameInfo]) -> SGInterp {
         // Get AOT module IR and parse it.
         let module = Module::from_bc();
+
+        // Initialise frames.
+        let mut frames = Vec::new();
+        for frame in activeframes {
+            let funcname = std::ffi::CStr::from_ptr(frame.fname);
+            let func = module.function(funcname.as_ptr());
+            let bb = func.bb(frame.bbidx);
+            let instr = bb.instruction(frame.instridx);
+            frames.push(Frame::new(instr));
+        }
         // Create and initialise stop gap interpreter.
-        let func = module.function(fname.as_ptr());
-        let bb = func.bb(bbidx);
-        let instr = bb.instruction(instridx);
+        let current_pc = frames.last().unwrap().pc;
         SGInterp {
             module,
-            frames: vec![Frame::new()],
-            pc: instr,
+            frames,
+            pc: current_pc,
             lastbb: None,
         }
     }
 
     /// Add a live variable and its value to the current frame.
-    pub fn var_init(&mut self, bbidx: usize, instridx: usize, fname: &CStr, val: u64) {
+    pub fn var_init(
+        &mut self,
+        bbidx: usize,
+        instridx: usize,
+        fname: &CStr,
+        sfidx: usize,
+        val: u64,
+    ) {
         let func = self.module.function(fname.as_ptr());
         let bb = func.bb(bbidx);
         let instr = bb.instruction(instridx);
-        let orgaot = unsafe { get_aot_original(&instr) };
+        let orgaot = if sfidx == 0 {
+            unsafe { get_aot_original(&instr) }
+        } else {
+            // Only the root stackframe has a control point call so there's no need to find
+            // "original" aot variables for the other frames.
```

Comment:
I'm not sure I understand this bit.

Reply:
I'm not sure what "original" means in this context. [Also s/aot/AOT/.]

Reply:
Try 3eff2493. You might need to re-read the description of `get_aot_var` to get some context.

Reply:
I think there's something implicit in the comment (but it's worth thinking explicitly about!): when we deoptimise in a trace, we know that we only have to recreate stackframes going back to the most recent control point in the trace (note: this might not be the control point that started the trace!). I think that might explain some of the confusion @vext01 and I have?

Reply:
While what you said is correct, it's not what this code is about. This is about initialising the correct variables in the stopgap interpreter.

Let's say we have this (simplified) AOT code:
```
%1 = alloca
store 1, %1
...
store %1 in YKCPStruct
control_point(YKCPStruct)
%2 = load YKCPStruct // this is the same as %1
store 2, %2
```

Now in the trace we only see:
```
%2 = load YKCPStruct
store 2, %2
```

When we guard fail, the stackmap will tell us that there's a variable %2 which has a value (e.g. `0xFF123`). So when we initialise the stopgap interpreter we do `Vars[%2] = 0xFF123`. However, the stopgap interpreter may now encounter the original variable `%1`, which wasn't initialise by the stackmap (since the trace has no knowledge of it). The workaround thus is, to check if `%2` is something we passed in via the `YKCPStruct`, and if so, also assign its value to the passed in value, in this case `%1`.

Reply:
What happens if `%1` and `%2` don't have the same value?

Reply:
They have to. Conceptually they are the same variable. It's just that the trace/stackmap doesn't know that as all it sees is the values passed in via the `control_point` call.

Reply:
Yes, it's the value of the same high-level variable before/after the CP.